zkevm-circuits/src/pi_circuit.rs

```diff
@@ -0,0 +1,978 @@
+                            rlc(tx.tx_sign_hash, self.randomness),
+                        ),
+                    ] {
+                        config.assign_tx_row(
+                            &mut region,
+                            offset,
+                            i + 1,
+                            *tag,
+                            0,
+                            *value,
+                            &mut raw_pi_vals,
+                        )?;
+                        offset += 1;
+                    }
+                }
+                // Tx Table CallData
+                let mut calldata_count = 0;
+                for (i, tx) in self.public_data.txs.iter().enumerate() {
+                    for (index, byte) in tx.call_data.0.iter().enumerate() {
+                        assert!(calldata_count < MAX_CALLDATA);
+                        config.assign_tx_row(
+                            &mut region,
+                            offset,
+                            i + 1,
+                            TxFieldTag::CallData,
+                            index,
+                            F::from(*byte as u64),
+                            &mut raw_pi_vals,
+                        )?;
+                        offset += 1;
+                        calldata_count += 1;
+                    }
+                }
+                for _ in calldata_count..MAX_CALLDATA {
+                    config.assign_tx_row(
+                        &mut region,
+                        offset,
+                        0, // tx_id
+                        TxFieldTag::CallData,
+                        0,
+                        F::zero(),
+                        &mut raw_pi_vals,
+                    )?;
+                    offset += 1;
+                }
+
+                // rpi_rlc and rand_rpi cols
+                let (rpi_rand, rpi_rlc) =
+                    config.assign_rlc_pi(&mut region, self.rand_rpi, raw_pi_vals)?;
+
+                Ok(vec![
+                    rpi_rand,
+                    rpi_rlc,
+                    chain_id,
+                    state_root,
+                    prev_block_hash,
+                ])
+            },
+        )?;
+
+        // Constrain raw_public_input cells to public inputs
+        for (i, pi_cell) in pi_cells.iter().enumerate() {
+            layouter.constrain_instance(pi_cell.cell(), config.pi, i)?;
+        }
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod pi_circuit_test {
+    use super::*;
+
+    use ethers_core::{
+        types::{NameOrAddress, TransactionRequest},
+        utils::keccak256,
+    };
+    use ethers_signers::{LocalWallet, Signer};
+    use halo2_proofs::{
+        dev::{MockProver, VerifyFailure},
+        pairing::bn256::Fr,
+    };
+    use pretty_assertions::assert_eq;
+    use rand::{CryptoRng, Rng, SeedableRng};
+    use rand_chacha::ChaCha20Rng;
+
+    // Copied from tx_circuit
+    fn rand_tx<R: Rng + CryptoRng>(mut rng: R, chain_id: u64) -> Transaction {
```

Comment:
We can move the `rand_tx` function from the tx_circuit to `zkevm-circuits/src/test_util.rs` which is only enabled under `test`: https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/4bdb2a2e70b6137ec0905bede725f5750fe4caf7/zkevm-circuits/src/lib.rs#L24-L25

So that we de-duplicate this function.